### ISSUE TITLE
[GSSoC 2026] [Feature] Add Real-Time Hand Gesture Recognition WebApp using MediaPipe and Deep Learning

### ISSUE THREAD

**Issue by Aarju247** [Niketkumardheeryan/ML-CaPsule#1328]
Hello @Niketkumardheeryan,

I would like to work on this issue under GSSoC'26.

### Proposed Implementation

I plan to develop a lightweight, real-time **Hand Gesture Recognition** system inspired by MediaPipe-based gesture recognition pipelines such as the kinivi hand-gesture-recognition-mediapipe architecture.

The project will use **MediaPipe Hands** for extracting 21 hand landmarks in real-time and a lightweight **MLP/Keypoint Classifier** for efficient gesture classification using normalized landmark coordinates instead of heavy image-based CNN models.

### Planned Features

* Real-time hand landmark detection using MediaPipe
* Lightweight gesture classification pipeline
* Recognition of gestures such as:

  * Open Palm
  * Closed Fist
  * Thumbs Up
  * Peace
  * OK Sign
* Live webcam prediction using OpenCV
* Optional Streamlit-based interactive UI
* CSV-based custom dataset generation for training new gestures
* Beginner-friendly and lightweight implementation

### Proposed Project Structure

```text
hand_gesture_recognition/
│
├── app.py
├── local_inference.py
│
├── model/
│   ├── keypoint_classifier/
│   └── keypoint_classifier.py
│
├── utils/
│   └── cv_utils.py
│
├── dataset/
│   ├── collect_data.py
│   └── keypoint.csv
│
├── notebooks/
│   └── model_training.ipynb
│
├── requirements.txt
└── README.md
```

### Tech Stack

* Python
* MediaPipe
* OpenCV
* NumPy
* Scikit-learn / PyTorch
* Streamlit

### Expected Deliverables

1. Real-time gesture recognition pipeline
2. Custom dataset collection script
3. Lightweight trained classification model
4. Webcam inference application
5. Training notebook with preprocessing and evaluation
6. Well-structured README.md and setup instructions

I have already explored similar MediaPipe-based implementations and would love to contribute this feature following the repository guidelines.

Please assign this issue to me under GSSoC'26.

Thank you!

**Comment by Niketkumardheeryan** [comment]
@Aarju247 expecting Ui will be there and you can gestures to control a particular functionality

**Comment by Aarju247** [comment]
Hello @Niketkumardheeryan, 

Yes, absolutely! I will definitely build a clean, interactive **Streamlit UI** for the application. 

To meet your expectations, I will map the recognized hand gestures to trigger specific functionalities within the app—for example:
* **Open Palm:** Pause/Play a media element or simulation.
* **Thumbs Up / Thumbs Down:** Toggle application themes (Light/Dark mode) or navigate dashboard tabs.
* **Fist:** Clear or reset the current application state.

I'll ensure the UI is fully functional and interactive. Looking forward to building this!

**Comment by couder-04** [comment]
Hi @Niketkumardheeryan ,

I would like to work on this issue under GSSoC '26.

## Proposed Approach

I plan to develop a real-time hand gesture recognition web application using MediaPipe and Deep Learning techniques for accurate gesture detection and classification through webcam input.

The application will detect hand landmarks in real time using MediaPipe Hands and perform gesture classification using a trained deep learning model. The project will focus on low-latency inference, responsive UI interaction, and robust gesture recognition performance.

## Planned Enhancements

- Real-time hand landmark detection using MediaPipe
- Deep learning-based gesture classification pipeline
- Webcam-based live gesture recognition interface
- Support for multiple hand gestures
- Real-time prediction display with confidence scores
- Clean and responsive web application UI
- Efficient preprocessing and inference pipeline
- Modular and maintainable project structure

## Features to be Included

- Hand tracking and landmark extraction
- Gesture dataset preprocessing pipeline
- Model training and evaluation
- Live prediction visualization
- Gesture confidence visualization
- Error handling for missing/partial hand detections
- Scalable architecture for adding new gestures later

## Goals

- Build an accurate and real-time gesture recognition system
- Improve user interaction through computer vision
- Create an educational and beginner-friendly AI project
- Demonstrate practical usage of MediaPipe and Deep Learning
- Ensure maintainability and scalability of the codebase

## Tech Stack

- Python
- MediaPipe
- TensorFlow / Keras
- OpenCV
- NumPy
- Flask / Streamlit
- Matplotlib (optional for analysis)

## Files Likely to be Modified

- app.py
- model_training.py
- gesture_detector.py
- utils/preprocessing.py
- templates/
- static/
- README.md

Name: Parth Tomar  
GitHub: https://github.com/couder-04  
Email: parth04pt@gmail.com  

Kindly assign this issue to me. I would love to contribute to this project and submit a quality PR.

— Parth Tomar (@couder-04)  
GSSoC 2026 Contributor | AI/ML Track
